### PR TITLE
Support custom labels for the HiveMQ Platform

### PR DESCRIPTION
**Pull request by Donnerbart**
https://hivemq.kanbanize.com/ctrl_board/22/cards/22802/details/

Example CR when all annotations and labels are set:
```yaml
---
# Source: hivemq-platform/templates/hivemq-custom-resource.yml
apiVersion: hivemq.com/v1
kind: HiveMQPlatform
metadata:
  name: "my-platform"
  annotations:
    global-annotation-key-1: value # <--- not configurable via Helm chart
    global-annotation-key-2: value # <--- not configurable via Helm chart
  labels:
    app.kubernetes.io/name: "hivemq-platform"
    app.kubernetes.io/instance: "my-platform"
    app.kubernetes.io/version: "4.30.0"
    global-label-key-1: value # <--- not configurable via Helm chart
    global-label-key-2: value # <--- not configurable via Helm chart
spec:
  configMapName: hivemq-configuration-my-platform
  logLevel: INFO
  operatorRestApiPort: 7979
  healthApiPort: 8889
  metricsPort: 9399
  metricsPath: /
  statefulSet:
    spec:
      replicas: 2
      template:
        metadata:
          annotations:
            pod-annotation-key-1: value # <--- NODES
            pod-annotation-key-2: value # <--- NODES
          labels:
            pod-label-key-1: value # <--- NODES
            pod-label-key-2: value # <--- NODES
        spec:
          containers:
            - name: hivemq
              image: "docker.io/hivemq/hivemq4:4.30.0"
              imagePullPolicy: IfNotPresent
              ports:
                - name: mqtt-1883
                  containerPort: 1883
                - name: cc-8080
                  containerPort: 8080
                  # Metric container ports are not configurable right now
                - name: metrics-9399
                  containerPort: 9399
              resources:
                limits:
                  cpu: "1024m"
                  memory: "2048M"
                requests:
                  cpu: "1024m"
                  memory: "2048M"
  services:
    - metadata:
        name: hivemq-my-platform-mqtt-1883
        annotations:
          service-annotation-key-1: value # <--- SERVICES
          service-annotation-key-2: value # <--- SERVICES
        labels:
          service-label-key-1: value # <--- SERVICES
          service-label-key-2: value # <--- SERVICES
      spec:
        ports:
          - name: mqtt-1883
            targetPort: mqtt-1883
            port: 1883
    - metadata:
        name: hivemq-my-platform-cc-8080
      spec:
        sessionAffinity: ClientIP
        ports:
          - name: cc-8080
            targetPort: cc-8080
            port: 8080
    # Metric service ports are not configurable right now
    - metadata:
        name: hivemq-my-platform-metrics-9399
      spec:
        ports:
          - name: metrics-9399
            targetPort: metrics-9399
            port: 9399
  extensions:
    - id: hivemq-allow-all-extension
      enabled: true
      supportsHotReload: false
      extensionUri: "preinstalled"
```

### CODE REVIEW

File: charts/hivemq-platform/templates/hivemq-custom-resource.yml [hunk 1]
```diff
@@ -7,11 +7,18 @@ apiVersion: hivemq.com/v1
 kind: HiveMQPlatform
 metadata:
   name: "{{ .Release.Name }}"
+  {{- if and (typeIsLike "map[string]interface {}" .Values.platform.annotations) (.Values.platform.annotations) }}
+  annotations:
+  {{- .Values.platform.annotations | toYaml | nindent 4 }}
+  {{- end }}
```
Comment (afalhambra-hivemq) [inline charts/hivemq-platform/templates/hivemq-custom-resource.yml:13]: checking the type is not really needed, as this validation is done by the JSON schema. Also, you could just simplify this expression by using a `with` statement (this is just a personal preference, `if` is also fine).

```suggestion
  {{- with .Values.platform.annotations }}
  annotations:
  {{- toYaml . | nindent 4 }}
  {{- end }}
```

File: charts/hivemq-platform/templates/hivemq-custom-resource.yml [hunk 2]
```diff
@@ -7,11 +7,18 @@ apiVersion: hivemq.com/v1
...
   labels:
     {{- include "hivemq-platform.labels" . | nindent 4 }}
     {{- if and (ne .Values.operator.selector "nil") (.Values.operator.selector) }}
     operator: {{.Values.operator.selector}}
     {{- end }}
+    {{- if and (typeIsLike "map[string]interface {}" .Values.platform.labels) (.Values.platform.labels) }}
```
Comment (afalhambra-hivemq) [inline charts/hivemq-platform/templates/hivemq-custom-resource.yml:19]: same comment regarding types check

File: charts/hivemq-platform/templates/hivemq-custom-resource.yml [hunk 3]
```diff
@@ -24,10 +31,16 @@ spec:
     spec:
       replicas: {{.Values.nodes.replicaCount}}
       template:
-        {{- if and (typeIsLike "map[string]interface {}" .Values.nodes.annotations) (.Values.nodes.annotations) }}
+        {{- if or (and (typeIsLike "map[string]interface {}" .Values.nodes.annotations) (.Values.nodes.annotations)) (and (typeIsLike "map[string]interface {}" .Values.nodes.labels) (.Values.nodes.labels)) }}
```
Comment (afalhambra-hivemq) [inline charts/hivemq-platform/templates/hivemq-custom-resource.yml:34]: same comment regarding types check

File: charts/hivemq-platform/templates/hivemq-custom-resource.yml [hunk 4]
```diff
@@ -211,6 +224,10 @@ spec:
         annotations:
           {{- $val.annotations | toYaml | nindent 10 }}
         {{- end }}
+        {{- if and (typeIsLike "map[string]interface {}" $val.labels) ($val.labels) }}
+        labels:
+          {{- $val.labels | toYaml | nindent 10 }}
+        {{- end }}
```
Comment (afalhambra-hivemq) [inline charts/hivemq-platform/templates/hivemq-custom-resource.yml:230]: same comment regarding types check

File: charts/hivemq-platform/values.yaml
Comment (afalhambra-hivemq) [inline charts/hivemq-platform/values.yaml:1]: would it make sense to include an IT where we set different values for the `.Values.platform.annotations/labels`, `.Values.nodes.annotations/labels` and `.Values.services.annotations/labels`?

File: charts/hivemq-platform/tests/hivemq_services_test.yaml [hunk 5]
```diff
@@ -104,6 +104,58 @@ tests:
...
 
+  - it: with additional labels
+    set:
+      services:
+        - type: mqtt
+          exposed: true
+          containerPort: 1884
+          labels:
+            test-label-key: test-label-value
+    asserts:
```
Comment (afalhambra-hivemq) [inline charts/hivemq-platform/tests/hivemq_services_test.yaml:115]: should we assert it doesn't contain any annotation?

File: charts/hivemq-platform/tests/hivemq_services_test.yaml [hunk 6]
```diff
@@ -104,6 +104,58 @@ tests:
                   targetPort: mqtt-1884
```
Comment (afalhambra-hivemq) [inline charts/hivemq-platform/tests/hivemq_services_test.yaml:104]: should we update this test to make sure it does not contain any label?

File: charts/hivemq-platform-operator/tests/hivemq_operator_deployment_test.yaml [hunk 7]
```diff
@@ -374,3 +374,31 @@ tests:
...
+          path: spec.template.spec.containers[0]
+      - exists:
+          path: spec.template.spec.containers[0].env
+      - contains:
+          path: spec.template.spec.containers[0].env
+          content:
+            name: hivemq.platform.operator.statefulset.trigger-rolling-restart-on-template-metadata-change
+            value: "true"
+
+  - it: with platform statefulset trigger-rolling-restart-on-template-metadata-change disabled
```
[the remark this answers is not in the document]
Comment (Donnerbart) [reply charts/hivemq-platform-operator/tests/hivemq_operator_deployment_test.yaml:392]: Once we have finally decided what the default value should be, we could adapt the according test case to be the default one and remove the `set` part from it. WDYT?

File: charts/hivemq-platform/tests/hivemq_services_test.yaml (hunk 5 above)
Comment (Donnerbart) [reply charts/hivemq-platform/tests/hivemq_services_test.yaml:115]: That should already be tested using `content`. When I add an annotation to the setup, the assertion fails:
```
 FAIL  test HiveMQ Platform services    charts/hivemq-platform/tests/hivemq_services_test.yaml
        - with additional labels

                - asserts[1] `contains` fail
                        Template:       hivemq-platform/templates/hivemq-custom-resource.yml
                        DocumentIndex:  0
                        Path:   spec.services
                        Expected to contain:
                                - metadata:
                                    labels:
                                      test-label-key: test-label-value
                                    name: hivemq-test-hivemq-platform-mqtt-1884
                                  spec:
                                    ports:
                                      - name: mqtt-1884
                                        port: 1884
                                        targetPort: mqtt-1884
                        Actual:
                                - metadata:
                                    annotations:
                                      foo: bar
                                    labels:
                                      test-label-key: test-label-value
                                    name: hivemq-test-hivemq-platform-mqtt-1884
                                  spec:
                                    ports:
                                      - name: mqtt-1884
                                        port: 1884
                                        targetPort: mqtt-1884
                                - metadata:
                                    name: hivemq-test-hivemq-platform-metrics-9399
                                  spec:
                                    ports:
                                      - name: metrics-9399
                                        port: 9399
                                        targetPort: metrics-9399
```

File: charts/hivemq-platform/tests/hivemq_services_test.yaml (hunk 6 above)
Comment (Donnerbart) [reply charts/hivemq-platform/tests/hivemq_services_test.yaml:104]: Same here, this is already asserted by just expecting a label using `content`.

File: charts/hivemq-platform/values.yaml
Comment (Donnerbart) [reply charts/hivemq-platform/values.yaml:1]: I've changed `hivemq_platform_test.yaml` to use `annotation-value-x` and `label-value-x`, so we can be sure the values are taken from the correct source. I also added a second annotation/label to the tests in `hivemq_services_test.yaml`.

Is that what you meant or are you suggesting another test case?

File: charts/hivemq-platform-operator/tests/hivemq_operator_deployment_test.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (afalhambra-hivemq) [reply charts/hivemq-platform-operator/tests/hivemq_operator_deployment_test.yaml:392]: Makes sense. Thanks

File: charts/hivemq-platform/tests/hivemq_services_test.yaml (hunk 5 above)
Comment (afalhambra-hivemq) [reply charts/hivemq-platform/tests/hivemq_services_test.yaml:118]: Sorry, I missed the use of `content`. My bad

File: charts/hivemq-platform/values.yaml
Comment (afalhambra-hivemq) [reply charts/hivemq-platform/values.yaml:1]: I was more thinking about having an integration test in place using custom values for the annotations and labels. As we don't have any here in that regard.
Though this is covered by a migration IT we have in the operator repo.
Not sure, WDYT?

**Review by afalhambra-hivemq — APPROVED** [review]
Nice alignment and improvement of the labels and annotations!

**Pull request merged**